src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java

```diff
@@ -44,69 +48,222 @@
+        this.testMode = testMode;
+    }
+
     /**
      *
-     * Register data collector with this service.
-     * The registered collectors will be called to provide heartbeat data.
+     * Register data collector with this service and start the schedule.
+     * The registered collector will be called to provide heartbeat data at the scheduled interval.
+     * Each collector registered via this method must have a unique collector id.
      *
      * @param collector collector to register
      */
     @Override
-    public void registerCollector(HBBaseDataCollector collector)
+    public void registerCollector(final HBBaseDataCollector collector)
+    {
+        // Check collector ID is not null
+        if (collector.getCollectorId() == null)
+        {
+            logger.error("HeartBeat did not registered collector, collector ID must not be null.");
+            return;
+        }
+
+        // Check collector with the same ID does't already exist
+        for (HBBaseDataCollector col : collectors)
+        {
+            if(collector.getCollectorId().equals(col.getCollectorId()))
+            {
+                logger.error("HeartBeat did not registered collector, ID must be unique. ID: " + collector.getCollectorId());
+                return;
+            }
+        }
+
+        // Schedule collector job and add collector to list of registered collectors
+        try
+        {
+            scheduleCollector(collector);
+            collectors.add(collector);
+
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("HeartBeat registered collector: " + collectorInfo(collector));
+            }
+        }
+        catch (Exception e)
+        {
+            logger.error("HeartBeat did not registered collector: " + collectorInfo(collector), e);
+        }
+    }
+
+    @Override
+    public boolean isEnabledByDefault()
+    {
+        return defaultHbState;
+    }
+
+    /**
+     * Start or stop the HertBeat jobs for all registered collectors depending on whether the heartbeat is enabled or not
+     */
+    private synchronized void scheduleCollector(final HBBaseDataCollector collector) throws ParseException, SchedulerException
+    {
+        final String jobName = "heartbeat-" + collector.getCollectorId();
+        final String triggerName = jobName + "-Trigger";
+
+        if (this.enabled)
+        {
+            scheduleJob(jobName, triggerName, collector);
+        }
+        else
+        {
+            unscheduleJob(triggerName, collector);
+        }
+
+    }
+
+    private void scheduleJob(final String jobName, final String triggerName, final HBBaseDataCollector collector) throws ParseException, SchedulerException
+    {
+        final JobDetail jobDetail = new JobDetail(jobName, Scheduler.DEFAULT_GROUP, HeartBeatJob.class);
+        final String cronExpression = testMode ? testCronExpression : collector.getCronExpression();
+        jobDetail.getJobDataMap().put("collector", collector);
```

Comment:
The keys in the Map ("collector", "hbDataSenderService" and "jobLockService") should be defined as ```public static final String``` fields on the HeartBeatJob class, and then referenced from the HBDataCollectorServiceImpl.java. Otherwise, it is easy to forget one or introduce a typo when changing names.